src/pool.rs

```diff
@@ -20,52 +22,129 @@ const MAX_IDLE_CONNECTIONS: usize = 100;
     // pool reaches MAX_IDLE_CONNECTIONS. The corresponding PoolKeys for
     // recently used Streams are added to the back of the queue;
     // old streams are removed from the front.
     lru: VecDeque<PoolKey>,
+    max_idle_hosts: usize,
+    max_idle_connections_per_host: usize,
 }
 
 impl ConnectionPool {
     pub fn new() -> Self {
         ConnectionPool {
+            max_idle_hosts: DEFAULT_MAX_IDLE_HOSTS,
+            max_idle_connections_per_host: DEFAULT_MAX_IDLE_CONNECTIONS_PER_HOST,
             ..Default::default()
         }
     }
 
+    pub fn set_max_idle_hosts(&mut self, max_hosts: usize) {
+        if self.max_idle_hosts == max_hosts {
+            return;
+        }
+        self.max_idle_hosts = max_hosts;
+
+        if max_hosts == 0 {
+            // Clear the connection pool, caching is disabled.
+            self.lru.clear();
+            self.recycle.clear();
+            return;
+        }
+
+        // Remove any extra hosts if the number was decreased.
+        while self.lru.len() > max_hosts {
+            let key = self.lru.pop_front().unwrap();
+            let removed = self.recycle.remove(&key);
+            assert!(
+                removed.is_some(),
+                "invariant failed: key existed in lru but not in recycle"
+            );
+        }
+    }
+
+    pub fn set_max_idle_connections_per_host(&mut self, max_connections: usize) {
+        if self.max_idle_connections_per_host == max_connections {
+            return;
+        }
+        self.max_idle_connections_per_host = max_connections;
+
+        if max_connections == 0 {
+            // Clear the connection pool, caching is disabled.
+            self.lru.clear();
+            self.recycle.clear();
+            return;
+        }
+
+        // Remove any extra streams if the number was decreased.
+        for val in self.recycle.values_mut() {
+            while val.len() > max_connections {
+                val.pop_front();
+            }
+        }
+    }
+
     /// How the unit::connect tries to get a pooled connection.
     pub fn try_get_connection(&mut self, url: &Url, proxy: &Option<Proxy>) -> Option<Stream> {
         let key = PoolKey::new(url, proxy);
-        self.remove(&key)
+        self.remove(key)
     }
 
-    fn remove(&mut self, key: &PoolKey) -> Option<Stream> {
-        if !self.recycle.contains_key(&key) {
-            return None;
+    fn remove(&mut self, key: PoolKey) -> Option<Stream> {
+        match self.recycle.entry(key.clone()) {
+            Entry::Occupied(mut occupied_entry) => {
+                let streams = occupied_entry.get_mut();
+                let stream = streams.pop_front();
+                assert!(
+                    stream.is_some(),
+                    "key existed in recycle but no streams available"
+                );
+                if streams.len() == 0 {
+                    occupied_entry.remove_entry();
+                    self.lru.retain(|x| x != &key);
+                }
```

Comment:
Wonder if the `key` should be shifted to the back of `lru` here? If the `stream` we use failed, it won't be reinserted in the pool, and I wonder if the order should be changed anyway?

Reply:
Sorry, I don't quite understand! `remove_entry` removes the key/now empty VecDeque from `self.recycle` when the last stream has been removed, and then `retain` removes `key` from `lru` as well. Do you mean keep the key in `lru` even when there are no streams of that key remaining in `recycle`? If the used `stream` fails, then I guess for the same `key` to be reinserted, it would mean the user created a new connection, or another in-use connection was returned. In both cases, `add` also moves the key to the back of `lru` at the moment.

Reply:
What I mean is in the case where `stream.len() > 0`. I.e. we have two connections for some host, we take one out and one remains – maybe we should shift the position of the remaining one?

Reply:
Ah, got it. That makes sense. Perhaps it should be shifted to the end of the `lru` since the host becomes the most recently used.

Reply:
Yeah. I think so.

Reply:
Please take another look. Because of the changes jsha suggested, I had to update this logic.